File: level2/ex35.js
function solution(arr) {
  // 초기 0와 1의 개수를 반환하는 함수
  function initialValues(arr, n) {
    // 0의 개수
    let count_zero = 0;
    // 1의 개수
    let count_one = 0;

    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        arr[i][j] === 0 ? count_zero++ : count_one++;
      }
    }

    return [count_zero, count_one];
  }

  // 배열 내 모든 요소가 같은 값인 경우 그 값과 개수를 반환하는 함수
  function sameValues(arr, n, startRow = 0, startColumn = 0) {
    // 같은 값인지 확인하기 위한 기준 값
    const value = arr[startRow][startColumn];

    for (let i = startRow; i < startRow + n; i++) {
      for (let j = startColumn; j < startColumn + n; j++) {
        // 다른 값이 존재하는 경우
        if (arr[i][j] !== value) {
          return [-1, -1];
        }
      }
    }

    return [value, n * n];
  }

  // 쿼드 압축 실행 함수
  function quardCompress(arr, n, startRow = 0, startColumn = 0) {
    // 현재 영역에서 모든 요소가 같은 값인지 확인
    const [value, count] = sameValues(arr, n, startRow, startColumn);

    // 0과 1의 개수 갱신
    if (value === 0) {
      count_zero -= count - 1;
    } else if (value === 1) {
      count_one -= count - 1;
    }

    // n 값 갱신
    n /= 2;

    // 더 이상 압축할 필요가 없는 경우
    if (n === 1 || value !== -1) {
      return;
    }

    // 좌측 상단 영역
    quardCompress(arr, n, startRow, startColumn);
    // 좌측 하단 영역
    quardCompress(arr, n, startRow + n, startColumn);
    // 우측 상단 영역
    quardCompress(arr, n, startRow, startColumn + n);
    // 우측 하단 영역
    quardCompress(arr, n, startRow + n, startColumn + n);
  }

  const n = arr.length;
  let [count_zero, count_one] = initialValues(arr, n);

  // 쿼드 압축 실행
  quardCompress(arr, n, 0, 0);

  return [count_zero, count_one];
}

File: level2/ex35.py
def solution(arr):
    # 주어진 2차원 리스트 내 0과 1의 개수를 반환
    def get_count(n, arr):
        # 0과 1의 개수
        count = [0, 0]
        
        for i in range(n):
            for j in range(n):
                value = arr[i][j]
                count[value] += 1
        
        return count
    
    # 주어진 2차원 리스트 내 (i, j)부터 시작하여 크기가 m x m인 2차원 리스트를 반환
    def slice(arr, i, j, m):
        return [[arr[a][b] for b in range(j, j+m)] for a in range(i, i+m)]
    
    
    # 주어진 2차원 리스트를 4개로 나누어 압축을 시도
    def compress(arr):
        # 2차원 리스트의 크기
        n = len(arr)
        # 현재 0과 1의 개수
        count = get_count(n, arr)
        
        # 0 또는 1로만 이루어졌을 경우 
        for i in range(len(count)):
            if count[i] == n*n:
                answer[i] -= n*n-1
                return 

        # 4개로 나눴을 때 2차원 리스트의 크기
        m = n // 2
        # 4개로 나누어 압축을 시도
        compress(slice(arr, 0, 0, m))
        compress(slice(arr, 0, m, m))
        compress(slice(arr, m, 0, m))
        compress(slice(arr, m, m, m))

    n = len(arr)
    answer = get_count(n, arr)
    compress(arr)
    
    return answer